### ISSUE TITLE
CloudWatch datasource returns null values in time series gaps

### ISSUE THREAD

**Issue by visit1985** [grafana/grafana#42607]
**What happened**:

Whenever we do a CloudWatch Metrics query on a Metric with a gap of at least 1 data point in the returned time series for the given time range, the datasource API returns a null value for the first missing data point in the gap.

<img src="https://user-images.githubusercontent.com/2237222/144282221-2b11ba82-493d-47d1-8a91-cd122fab04ec.png" width="300">

<img src="https://user-images.githubusercontent.com/2237222/144281883-87816cba-ec5b-4afd-96ca-feea610989e5.png" width="554">

This null value is not returned by the CloudWatch API, if we do the same Metrics query using aws-cli.

```
aws cloudwatch get-metric-statistics \
--namespace AWS/Lambda \
--metric-name Invocations \
--dimensions Name=FunctionName,Value=my-function \
--statistics SampleCount \
--period 60 \
--start-time 2021-12-01T16:14:00Z \
--end-time 2021-12-01T16:19:00Z
```

This issue becomes serious, when we try to `Reduce` this time series using the new unified alerting, which returns `NaN` in this case. There is currently no way to transform or filter null values out.

**What you expected to happen**:

No null values to be added to the time series.

**Anything else we need to know?**:

As a workaround for unified alerting one can set the `Period` equivalent to the time range of the query. This will return a time series with one value only and thus no gap which can lead to additional null values.

**Environment**:
- Grafana version: v8.2.5
- Data source type & version: CloudWatch
- OS Grafana is installed on: Docker on Linux
- User OS & Browser: Chrome

**Comment by sunker** [comment]
Thanks for creating this issue. 

The CloudWatch data source in grafana is not using the GetMetricStatistics api - it's using the [GetMetricData](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricData.html) api. Can you please verify that this api is also not returning null values?

**Comment by visit1985** [comment]
@sunker, I can confirm that GetMetricsData does not return null values either.

<details>
  <summary>Query inspector</summary>

  ![image](https://user-images.githubusercontent.com/2237222/144640530-192b2023-b7a8-4b3f-a209-d1e70a563a09.png)
  
  ```json
  {
    "request": {
      "method": "POST",
      "url": "api/ds/query",
      "data": {
        "from": "1638375300000",
        "to": "1638375540000",
        "queries": [
          {
            "intervalMs": 200,
            "maxDataPoints": 1102,
            "datasourceId": 1,
            "type": "timeSeriesQuery",
            "namespace": "AWS/Lambda",
            "metricName": "Invocations",
            "expression": "",
            "dimensions": {
              "FunctionName": [
                "my-function"
              ]
            },
            "region": "eu-central-1",
            "id": "m1",
            "alias": "",
            "statistic": "SampleCount",
            "period": "60",
            "refId": "A",
            "matchExact": true,
            "datasource": "CloudWatch"
          }
        ]
      },
      "hideFromInspector": false
    },
    "response": {
      "results": {
        "A": {
          "frames": [
            {
              "schema": {
                "name": "Invocations_SampleCount",
                "refId": "A",
                "meta": {
                  "custom": {
                    "id": "m1",
                    "period": 60
                  }
                },
                "fields": [
                  {
                    "name": "Time",
                    "type": "time",
                    "typeInfo": {
                      "frame": "time.Time",
                      "nullable": true
                    }
                  },
                  {
                    "name": "Value",
                    "type": "number",
                    "typeInfo": {
                      "frame": "float64",
                      "nullable": true
                    },
                    "labels": {
                      "FunctionName": "my-function"
                    },
                    "config": {
                      "displayNameFromDS": "Invocations_SampleCount",
                      "links": [
                        {
                          "title": "View in CloudWatch console",
                          "targetBlank": true,
                          "url": "REDACTED"
                        }
                      ]
                    }
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    1638375300000,
                    1638375360000,
                    1638375420000,
                    1638375480000
                  ],
                  [
                    2,
                    1,
                    null,
                    1
                  ]
                ]
              }
            }
          ],
          "refId": "A"
        }
      }
    }
  }
  ```
</details>

<details>
  <summary>GetMetricData</summary>
  
  ```
  aws cloudwatch get-metric-data \
  --start-time 2021-12-01T16:14:00Z \
  --end-time 2021-12-01T16:19:00Z \
  --metric-data-queries '[
    {
      "Id": "m1",
      "MetricStat": {
        "Metric": {
          "Namespace": "AWS/Lambda",
          "MetricName": "Invocations",
          "Dimensions": [
            {
              "Name": "FunctionName",
              "Value": "my-function"
            }
          ]
        },
        "Period": 60,
        "Stat": "SampleCount",
        "Unit": "Count"
      }
    }
  ]'
  ```
  
  ```json
  {
      "MetricDataResults": [
          {
              "Id": "m1",
              "Label": "Invocations",
              "Timestamps": [
                  "2021-12-01T16:18:00Z",
                  "2021-12-01T16:16:00Z",
                  "2021-12-01T16:15:00Z"
              ],
              "Values": [
                  1.0,
                  1.0,
                  2.0
              ],
              "StatusCode": "Complete"
          }
      ],
      "Messages": []
  }
  ```
</details>

<details>
  <summary>GetMetricStatistics</summary>
  
  ```
  aws cloudwatch get-metric-statistics \
  --namespace AWS/Lambda \
  --metric-name Invocations \
  --dimensions Name=FunctionName,Value=my-function \
  --statistics SampleCount \
  --period 60 \
  --start-time 2021-12-01T16:14:00Z \
  --end-time 2021-12-01T16:19:00Z
  ```
  
  ```json
  {
      "Label": "Invocations",
      "Datapoints": [
          {
              "Timestamp": "2021-12-01T16:15:00Z",
              "SampleCount": 2.0,
              "Unit": "Count"
          },
          {
              "Timestamp": "2021-12-01T16:18:00Z",
              "SampleCount": 1.0,
              "Unit": "Count"
          },
          {
              "Timestamp": "2021-12-01T16:16:00Z",
              "SampleCount": 1.0,
              "Unit": "Count"
          }
      ]
  }
  ```
</details>

**Comment by fridgepoet** [comment]
Thanks for verifying and providing those API responses!

I'm able to reproduce the behavior from latest Grafana main exactly as described: 
* I created a simple lambda function
* I set the period to 60
* I invoke the lambda every minute
* As soon as I miss one period duration, I see a `null` gap in the response

The problem was that the results were prepared without any "gaps" in mind. I've prepared a [fix](https://github.com/grafana/grafana/pull/43507), but I'd like to discuss with the team first to ensure this behavior was not intentional.
